File: 12000/800/12865_평범한_배낭/BJ_12865_평범한_배낭.py
# https://www.acmicpc.net/problem/12865

import sys

readline = lambda: sys.stdin.readline().strip()

N, K = map(int, readline().split())
items = [tuple(map(int, readline().split())) for _ in range(N)]

dp = [[0] * (K + 1)]
visited = set([0])
for i, (w, v) in enumerate(items, start=1):
    dp.append([*dp[-1]])
    for w0 in [*visited]:
        if w0 + w <= K:
            visited.add(w0 + w)
            dp[i][w0 + w] = max(dp[i - 1][w0 + w], dp[i - 1][w0] + v)

print(max(dp[-1]))

# N, K = map(int, readline().split())
# items = [tuple(map(int, readline().split())) for _ in range(N)]

# dp = [[0] * (K + 1)]
# for w, v in items:
#     dp.append([0] * (K + 1))
#     for w0 in range(K + 1):
#         dp[-1][w0] = dp[-2][w0]
#         dp[-1][w0] = max(dp[-1][w0], dp[-2][w0 - w] + v) if 0 <= w0 - w else dp[-1][w0]

# print(max(dp[-1]))

File: 12000/800/12865_평범한_배낭/BJ_12865_평범한_배낭.cc
// https://www.acmicpc.net/problem/12865

#include <algorithm>
#include <cstdint>
#include <iostream>
#include <iterator>
#include <tuple>
#include <vector>

using namespace std;
using i32 = int;

int main(int argc, char const* argv[]) {
    auto in = istream_iterator<i32>{cin};
    ios::sync_with_stdio(false);
    cin.tie(nullptr);

    auto N = *in;
    auto K = *++in;
    auto items = vector<tuple<i32, i32>>(0);
    items.reserve(N);
    for (i32 i = 0; i < N; ++i) {
        auto const w = *++in;
        auto const v = *++in;
        if (w <= K) {
            items.emplace_back(w, v);
        }
    }

    auto dp = vector<i32>(K + 1);
    auto maxv = 0;
    for (auto[w, v] : items) {
        for (i32 i = K; 0 <= i; --i) {
            if (0 <= i - w) {
                if ((i - w == 0 || 0 < dp[i - w]) && dp[i] < dp[i - w] + v) {
                    dp[i] = dp[i - w] + v;
                    maxv = max(maxv, dp[i]);
                }
            }
        }
    }

    cout << maxv << endl;
    return 0;
}
